File: homework/Truong_Duong/Lesson_11/159_Longest_Substring_with_at_most_two_distinct_characters.js
// Solution for https://leetcode.com/problems/longest-substring-with-at-most-two-distinct-characters/

/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstringTwoDistinct = function (s) {
  let left = 0, right = 0, ans = 0;
  const N = s.length, countMap = {};

  let distinctCount = 1;
  countMap[s[0]] = 1;

  while (right < N) {
    if (distinctCount <= 2) {
      ans = Math.max(ans, right - left + 1);

      if (right + 1 < N) {
        right += 1;
        if (countMap[s[right]] >= 1) {
          countMap[s[right]] += 1;
        } else {
          countMap[s[right]] = 1;
          distinctCount += 1;
        }
      } else {
        break;
      }
    } else {
      countMap[s[left]] -= 1;
      if (countMap[s[left]] === 0) distinctCount -= 1;
      left += 1;
    }
  }

  return ans;
};

File: homework/thinhbka/lesson-7-twopointer-heap/159. Longest Substring with At Most Two Distinct Characters.py
class Solution:
    def lengthOfLongestSubstringTwoDistinct(self, s: str) -> int:
        l = 0
        ans = 0
        charSet = {}
        for r in range(len(s)):
            charSet[s[r]] = charSet.get(s[r],0)+1
            while len(charSet) == 3:
                charSet[s[l]]-=1
                if charSet[s[l]] == 0:
                    del charSet[s[l]]
                l+=1
            ans = max(ans,r-l+1)
        return ans